README.md
This is team project java GUI created by
- [Samuel Tolossa UGR/25454/14](https://github.com/tolossamuel) 
- [Abel Alemayehu UGR/25383/14](https://github.com/Zeordi)
- [Tinsae Teferi UGR/25753/14](https://github.com/tinsae316)
- [Bealprasim Demere UGR/25540 /14](https://github.com/prasimbaal)
- [Biruk Sitota UGR/25647/14](https://github.com/BileyX)

For temporary reviews of the system use Admin password 
- [username = "admin"](username)
- [password = "admin"](password)
you can access and add your student and Teacher to the system and check how the systems work

# Student Management System

The Student Management System is a simple Java Swing application that allows administrators to manage student and teacher records, and provides statistics on student and teacher data.

## Table of Contents

- [Features](#features)
- [Prerequisites](#prerequisites)
- [Installation](#installation)
- [Usage](#usage)
- [Contributing](#contributing)
- [License](#license)

## Student Management System
The Student Management System is a dynamic Java Swing application that empowers administrators, students, and teachers to efficiently handle academic records while providing insightful statistics about the educational environment. This beautifully organized system offers an array of features with intuitive functionality.

## Key Features
## Admin Panel
- [Login with Admin Credentials](StudentManagements.java)
- [Admin Dashboard](Admin.java)
- [Student and Teacher Management](Admin.java)
- [Grade Management](Admin.java)
- [Easy Navigation](Admin.java)
- [Features](Admin.java)
## The Admin Panel is your gateway to system management. Log in as an administrator with the username "admin" and password "admin". From here, you can manage student and teacher records, view statistics on the system's population, oversee student grades, and easily navigate between these essential functions.

## Student Panel

- [Student Login](StudentManagements.java)
- [Student Dashboard](Students.java)
- [Course Registration ](Students.java)
- [Grade Report ](Students.java)
- [Total Credit Hours ](Students.java)
- [Logout ](Students.java)
## The Student Panel provides a personalized space for students. After logging in, students can register for courses, check their grades, and monitor their total credit hours. A clean interface ensures a seamless experience.

## Teacher Panel

- [Teacher Login](StudentManagements.java)
- [Teacher Dashboard](Teacher.java)
- [Student List](Teacher.java)
- [Grade Management](Teacher.java)
- [Save Changes](Teacher.java)
- [Logout](Teacher.java)
## Teachers can log in to access the Teacher Panel, where they can view their students' records, manage and update grades to update grades just click on the grad row end write on in after that press enter and click save buttons, and save changes conveniently. This interactive platform simplifies academic management tasks for educators.

## Prerequisites
Before you embark on this academic adventure, ensure you meet these prerequisites:

Java Development Kit (JDK) installed on your machine.
A Java Integrated Development Environment (IDE) or code editor for development, such as IntelliJ IDEA, Eclipse, or Visual Studio Code.


History.txt
Samuel Tolossa	History	H123	3
Abel Alemayeyu	History	H123	3
Tinsae Teferi	History	H123	3
Biruk Stota	History	H123	3
Bealprasim Demere	History	H123	3


help.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;

public class help {
    private JFrame helpFrame = new JFrame("Help");

    help() {
        helpFrame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        helpFrame.setSize(600, 400);

        // Create a back button and add it to the top left corner
        JButton backButton = new JButton("Back");
        backButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                helpFrame.dispose();
                try {
                    MainClass.main(null);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        JPanel buttonPanel = new JPanel(new FlowLayout());
        buttonPanel.add(backButton);

        JPanel helpLabels = new JPanel();
        helpLabels.setLayout(new BoxLayout(helpLabels, BoxLayout.Y_AXIS)); // Vertical layout

        // Add your help content to the helpLabels panel (your Q&A JLabels)
		addQA(helpLabels, "Q0: How reviews the system?", "To reviews our system we have already temporary password and username for all (students, teachers, admin) <html><h3>for students username 'admin' and password also 'admin'</h3> <h3>for teachers the username is  'samy' password also 'samy'</h3> It is case sensitive this password and username not give full access of system it is just template access and to reviews how system work you access the template file only</html>");
        addQA(helpLabels, "Q1: How do I register for the system?", "Registration is done by the system administrator. They will provide you with a username and password. Please contact the administrator to get registered.");
        addQA(helpLabels, "Q2: What if I forget my password?", "If you forget your password, you can contact the system administrator. They can reset your password for you.");
        addQA(helpLabels, "Q3: I'm a student. How do I log in?", "Once registered, you can log in with the username and password given to you by the administrator. After logging in, you can access various features, including course registration and grade viewing.");
        addQA(helpLabels, "Q4: How do I register for courses?", "After logging in, you can register for courses that are scheduled for the semester. Use the registration feature in your student portal to select your desired courses.");
        addQA(helpLabels, "Q5: How can I view my grades?", "You can view your grades by logging in to your student portal. There, you will find an option to view your grade reports, which will display your test results and grades.");
        addQA(helpLabels, "Q6: I'm a teacher. How do I submit grades for my students?", "Teachers can submit grades for their assigned students through the system. Log in as a teacher, select the course and the student, and then submit the student's test results.");
        addQA(helpLabels, "Q7: Can I change my registered courses after registration?", "Once you've registered for a course, you may not be able to change it. Please check with the administrator or system guidelines for specific policies.");
        addQA(helpLabels, "Q8: How are grades calculated?", "Grades are calculated based on a predefined system that takes into account the test results. The system assigns grades based on specific score ranges, which you can view in your grade report.");

        JScrollPane scrollPane = new JScrollPane(helpLabels); // Make the content scrollable
        helpFrame.add(buttonPanel, BorderLayout.NORTH);
        helpFrame.add(scrollPane, BorderLayout.CENTER); // Add the scroll pane
	
        helpFrame.setVisible(true);
    }

    private void addQA(Container container, String question, String answer) {
        String wrappedAnswer = wrapText(answer, 80); // Wrap text into lines of max 80 characters
        container.add(new JLabel("<html><h2>" + question + "</h2>" + wrappedAnswer + "</html>"));
		
    }
	

    private String wrapText(String text, int maxLineLength) {
        StringBuilder wrappedText = new StringBuilder();
        String[] words = text.split(" ");
        StringBuilder line = new StringBuilder();

        for (String word : words) {
            if (line.length() + word.length() <= maxLineLength) {
                if (line.length() > 0) {
                    line.append(" ");
                }
                line.append(word);
            } else {
                wrappedText.append(line.toString()).append("<br>");
                line = new StringBuilder(word);
            }
        }

        if (line.length() > 0) {
            wrappedText.append(line.toString());
        }

        return wrappedText.toString();
    }

    protected void helpLogin(boolean states) throws IOException {
        helpFrame.setLocationRelativeTo(null);
        helpFrame.setVisible(states);
    }

}


about.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;

public class about {
    private JFrame aboutFrame = new JFrame("About");

    about() {
        aboutFrame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        aboutFrame.setSize(600, 400);  // Increased the width for horizontal scrolling

        // Create a back button and add it to the top left corner
        JButton backButton = new JButton("Back");
        backButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                aboutFrame.dispose();
                try {
                    MainClass.main(null);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        JPanel buttonPanel = new JPanel(new FlowLayout());
        buttonPanel.add(backButton);

        // Create a table with columns: Name, ID numbers, Professions
        String[] columnNames = {"<html><h2>Name</h2></html>", "<html><h2>ID numbers</h2></html>", "<html><h2>Professions</h2></html"};
        Object[][] data = {
            {"<html><h3>Samuel Tolossa</h3></html>", "<html><h3>UGR/25454/14</h3></html>", "<html><h3>Software Developers</h3></html>"},
	    {"<html><h3>Meron Bahiru</h3></html>", "<html><h3>UGR/25880/14 </h3></html>", "<html><h3>Software Developers</h3></html>"},
            {"<html><h3>Abel Alemayehu</h3></html>", "<html><h3>Ugr/25383/14</h3></html>", "<html><h3>Software Developers</h3></html>"},
            {"<html><h3>Tinsae Teferi</h3></html>", "<html><h3>UGR/25753/14</h3></html>", "<html><h3>Software Developers</h3></html>"},
            {"<html><h3>Bealprasim Demere</h3></html>", "<html><h3>UGR/25540/14</h3></html>", "<html><h3>Software Developers</h3></html>"},
            {"<html><h3>Biruk Sitota</h3></html>", "<html><h3>UGR/25647/14</h3></html>", "<html><h3>Software Developers</h3></html>"},
        };

        JTable table = new JTable(data, columnNames);
        JScrollPane tableScrollPane = new JScrollPane(table);

        // Create a JTextArea for "About Projects" text
        String aboutProjectsText = "This is a brief overview of our projects. We are a group of software developers working on various projects. Our team consists of talented individuals who are dedicated to creating high-quality software solutions.";
        aboutFrame.add(new JLabel(aboutProjectsText));
		JTextArea aboutProjectsTextArea = new JTextArea(5, 10);
        aboutProjectsTextArea.setText(aboutProjectsText);
        aboutProjectsTextArea.setLineWrap(true);
        aboutProjectsTextArea.setWrapStyleWord(true);

        JScrollPane textScrollPane = new JScrollPane(aboutProjectsTextArea);
        textScrollPane.setHorizontalScrollBarPolicy(JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);

        // Create a JPanel for the "About Projects" text
        JPanel textPanel = new JPanel(new BorderLayout());
        textPanel.add(textScrollPane);

        // Create a container panel for both the table and the "About Projects" text
        JPanel containerPanel = new JPanel(new BorderLayout());
        containerPanel.add(tableScrollPane, BorderLayout.NORTH);
        containerPanel.add(textPanel, BorderLayout.SOUTH);

        aboutFrame.add(buttonPanel, BorderLayout.NORTH);
        aboutFrame.add(containerPanel, BorderLayout.CENTER);
        aboutFrame.setVisible(true);
    }

    protected void aboutInfo(boolean states) throws IOException {
        aboutFrame.setLocationRelativeTo(null);
        aboutFrame.setVisible(states);
    }
}


Teacher.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;


public class Teacher {
    private String name;
    private Object[][] studentList;
	private String course;
    JFrame teacherStudents;
	
    Container contentPanel;

    Teacher(String name, String course) throws IOException {
		this.course = course;
		this.name = name;
		contentPanel = new JPanel();
		
        studentList = new Object[0][3];
        readFile(course);
        
        fremForTeacher();
    }

    private void fremForTeacher() throws IOException {
        teacherStudents = new JFrame("Teacher Students");
        teacherStudents.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
    	teacherStudents.setSize(600, 400);

        contentPanel.removeAll();

        JButton save = new JButton("Save Change");
        JButton logout = new JButton("Logout");

        JPanel buttonPanel = new JPanel(new FlowLayout());
        buttonPanel.add(logout);
        buttonPanel.add(save);
    

        JPanel studentListForm = new JPanel();
        studentListForm.setLayout(new BoxLayout(studentListForm, BoxLayout.Y_AXIS));

        studentListForm.add(new JLabel("Dashboard"));

        JLabel nameLabel = new JLabel("Name: Instructor " + name);
		JLabel nameLabel1 = new JLabel("Course " + course);
        nameLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        studentListForm.add(nameLabel);
		nameLabel1.setAlignmentX(Component.LEFT_ALIGNMENT);
        studentListForm.add(nameLabel1);

        JLabel semesterCoursesLabel = new JLabel("Your Students:");
        semesterCoursesLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        studentListForm.add(semesterCoursesLabel);

        String[] columnNames = {"Name Students", "Credit Hour", "result"};
        

        JTable courseTable = new JTable(studentList, columnNames);
        JScrollPane scrollPane = new JScrollPane(courseTable);
        scrollPane.setAlignmentX(Component.LEFT_ALIGNMENT);
        studentListForm.add(scrollPane);
		
        logout.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                teacherStudents.dispose();
                try {
                    MainClass.main(null);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });
		save.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                try {
					iterateTables(courseTable,course);
					
				} catch (IOException e1) {
					
					e1.printStackTrace();
				}
            }
        });
        contentPanel.setLayout(new BorderLayout());
        contentPanel.add(buttonPanel, BorderLayout.NORTH);
        contentPanel.add(studentListForm, BorderLayout.CENTER);

        teacherStudents.add(contentPanel);
        teacherStudents.revalidate();
        teacherStudents.repaint();
    }

    private void readFile(String course) throws IOException {
        File file = new File(course + "_Grad.txt");
        if (!file.exists()) {
            file.createNewFile();
        }
        BufferedReader readFile = new BufferedReader(new FileReader(file));
        String data;
        
        while ((data = readFile.readLine()) != null) {
			String[] dataRead = data.split("\t");
			
			// Ensure dataRead contains all three elements
			if (dataRead.length == 3) {
				// Create a new array to hold the data
				Object[] rowData = new Object[3];
				
				// Populate rowData with the data from dataRead
				for (int i = 0; i < 3; i++) {
					rowData[i] = dataRead[i];
				}
	
				// Add rowData to the studentList
				int currentLength = studentList.length;
				Object[][] newStudentList = new Object[currentLength + 1][];
				System.arraycopy(studentList, 0, newStudentList, 0, currentLength);
				newStudentList[currentLength] = rowData;
				studentList = newStudentList;
			}
		}
        
		
    }
	private void iterateTables( JTable courseTable,String course ) throws IOException{
		int rowCount = courseTable.getRowCount();
		int colCount = courseTable.getColumnCount();

		File rewriteFile = new File(course+"_Grad.txt");
		if (rewriteFile.exists()){
			rewriteFile.delete();
			rewriteFile.createNewFile();
		}
		// FileWriter editStudentMark = new FileWriter(rewriteFile);
		// editStudentMark.write("");
		// editStudentMark.close();
		FileWriter gradEdited = new FileWriter(rewriteFile,true);
		String data = "";
		for (int row = 0; row < rowCount; row++) {
			for (int col = 0; col < colCount; col++) {
				Object value = courseTable.getValueAt(row, col);
				data += value + "\t";
			}
			data += "\n";
			gradEdited.write(data);
			data = "";
			 // Move to the next row
		}
		gradEdited.close();
		studentList = new Object[0][];
		readFile(course);
	}
	protected void teacherAfterLogin(boolean states) throws IOException{
		
		fremForTeacher();
        teacherStudents.setLocationRelativeTo(null);
        teacherStudents.setVisible(states);
	}
}


Students.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.text.DecimalFormat;
import java.util.Arrays;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

class Students {
    private JFrame studentFrame = new JFrame("Student Management");
    private JPanel contentPanel = new JPanel();
    private JPanel registrationForm;
    private JLabel welcomeLabel;
    protected String name;

    private Object[][] myList;
    private Object[][] myGrad;

    Students() throws IOException {
        StudentManagements LoginName = new StudentManagements();
        name = LoginName.geter();
        myList = new Object[0][];
        myGrad = new Object[0][];
        this.showWelcomeMessage(name);
    }

    public void showRegistrationForm(String name) {
        contentPanel.removeAll();

        // Create a back button and add it to the top left corner
        JButton backButton = new JButton("Back");
        contentPanel.add(backButton, BorderLayout.NORTH);
        studentFrame.revalidate();
        studentFrame.repaint();

        // Create the registration form panel
        JPanel registrationFormPanel = new JPanel(new GridLayout(4, 2, 5, 5));

        // Add components to the registration form
        registrationFormPanel.add(new JLabel("Course field:"));
        // Create a combo box for the course field
        String[] courseOptions = {"Fundamental of Software Engineering", "DLD", "Object Oriented Programming", "History", "Psychology"};
        JComboBox<String> courseComboBox = new JComboBox<>(courseOptions);
        registrationFormPanel.add(courseComboBox);

        registrationFormPanel.add(new JLabel("Course code:"));
        JTextField courseCodeTextField = new JTextField(20);
        registrationFormPanel.add(courseCodeTextField);

        registrationFormPanel.add(new JLabel("Credit hours:"));
        JTextField creditHoursTextField = new JTextField(20);
        registrationFormPanel.add(creditHoursTextField);

        JButton saveButton = new JButton("Save");
        registrationFormPanel.add(saveButton);

        saveButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                String course = courseComboBox.getSelectedItem().toString();
                String creditHour = creditHoursTextField.getText();
                String courseCode = courseCodeTextField.getText();
                try {
                    if (course.trim().isEmpty() || creditHour.trim().isEmpty() || courseCode.trim().isEmpty()) {
                        JOptionPane.showMessageDialog(studentFrame, "Fill all spaces", "Error", JOptionPane.ERROR_MESSAGE);
                    } else if (isNotRegistered(name, course)) {
                        try {
                            savedData(name, course, creditHour, courseCode);
                            JOptionPane.showMessageDialog(studentFrame, "Saved successfully!!", "Success", JOptionPane.INFORMATION_MESSAGE);
                        } catch (NumberFormatException f) {
                            JOptionPane.showMessageDialog(studentFrame, "Invalid input of credit hour", "Error", JOptionPane.ERROR_MESSAGE);
                        }
                    } else {
                        JOptionPane.showMessageDialog(studentFrame, "This student is already registered for the selected course.", "Error", JOptionPane.ERROR_MESSAGE);
                    }
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        // Add an action listener for the back button
        backButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                try {
                    showWelcomeMessage(name);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });
        // Add the registration form panel to the content panel
        contentPanel.add(registrationFormPanel, BorderLayout.CENTER);

        studentFrame.revalidate();
        studentFrame.repaint();
    }

    public void showGradeReport(String name) throws IOException {
        contentPanel.removeAll();
        String[] columnNames = {"Course", "Credit Hours", "Result", "Grade"};
        String[] courseOptions = {"Fundamental of Software Engineering", "DLD", "Object Oriented Programming", "History", "Psychology"};

        for (int i = 0; i < courseOptions.length; i++) {
            readGrad(courseOptions[i], name);
        }

        // Create a back button and add it to the top left corner
        String totalGrad = CalculateGrade();
        totalGrad = String.format("<html><h3>Grade Report: %s</h3></html>", totalGrad);

        JLabel totalGradReport = new JLabel(totalGrad);
        totalGradReport.setHorizontalAlignment(SwingConstants.CENTER);
        totalGradReport.setVerticalAlignment(SwingConstants.BOTTOM);
        totalGradReport.setFont(totalGradReport.getFont().deriveFont(Font.BOLD));

        contentPanel.add(totalGradReport);

        JButton backButton = new JButton("Back");
        contentPanel.add(backButton, BorderLayout.NORTH);
        studentFrame.revalidate();
        studentFrame.repaint();

        JTable gradeTable = new JTable(myGrad, columnNames);
        JScrollPane scrollPane = new JScrollPane(gradeTable);
        contentPanel.add(scrollPane, BorderLayout.CENTER);

        backButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                try {
                    showWelcomeMessage(name);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        studentFrame.revalidate();
        studentFrame.repaint();
    }

    public void showWelcomeMessage(String name) throws IOException {
        studentFrame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        studentFrame.setSize(600, 400);
        contentPanel.removeAll();

        JButton registerButton = new JButton("Register");
        JButton Logout = new JButton("Logout");
        JButton gradeReportButton = new JButton("Grade Report");

        JPanel buttonPanel = new JPanel(new FlowLayout());
        buttonPanel.add(Logout);
        buttonPanel.add(registerButton);
        buttonPanel.add(gradeReportButton);

        registrationForm = new JPanel();
        registrationForm.setLayout(new BoxLayout(registrationForm, BoxLayout.Y_AXIS));

        registrationForm.add(new JLabel("Dashboard"));
        JLabel nameLabel = new JLabel("Name:  " + name);
        nameLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        registrationForm.add(nameLabel);

        JLabel semesterCoursesLabel = new JLabel("This Semester Courses:");
        semesterCoursesLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        registrationForm.add(semesterCoursesLabel);

        String[] columnNames = {"Course", "Course Code", "Credit Hours"};
String[] courseOptions = {"Fundamental of Software Engineering", "DLD", "Object Oriented Programming", "History", "Psychology"};
        
        for (int i = 0; i < courseOptions.length; i++) {
            registerCourse(courseOptions[i], name);
        }

        JTable courseTable = new JTable(myList, columnNames);
        JScrollPane scrollPane = new JScrollPane(courseTable);
        scrollPane.setAlignmentX(Component.LEFT_ALIGNMENT);
        registrationForm.add(scrollPane);

        int totalCreditHours = 0;
        for (int i = 0; i < myList.length; i++) {
            totalCreditHours += Integer.parseInt((String) myList[i][2]);
        }

        JLabel totalCreditHoursLabel = new JLabel("Total Credit Hours: " + totalCreditHours);
        totalCreditHoursLabel.setAlignmentX(Component.LEFT_ALIGNMENT);
        registrationForm.add(totalCreditHoursLabel);

        welcomeLabel = new JLabel("Welcome!");
        welcomeLabel.setHorizontalAlignment(SwingConstants.CENTER);

        registerButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                showRegistrationForm(name);
            }
        });

        Logout.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                studentFrame.dispose();
                try {
                    MainClass.main(null);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        gradeReportButton.addActionListener(new ActionListener() {
            public void actionPerformed(ActionEvent e) {
                try {
                    showGradeReport(name);
                } catch (IOException e1) {
                    e1.printStackTrace();
                }
            }
        });

        contentPanel.setLayout(new BorderLayout());
        contentPanel.add(buttonPanel, BorderLayout.NORTH);
        contentPanel.add(registrationForm, BorderLayout.CENTER);

        studentFrame.add(contentPanel);
        studentFrame.revalidate();
        studentFrame.repaint();
    }

    public String studentAfterLogin(boolean states, String name) throws IOException {
        this.showWelcomeMessage(name);
        studentFrame.setLocationRelativeTo(null);
        studentFrame.setVisible(states);
        return name;
    }

    void savedData(String name, String course, String creditHour, String courseCode) throws IOException {
        File studentFile = new File(course + ".txt");
        if (!studentFile.exists()) {
            studentFile.createNewFile();
        }
        FileWriter fileWriter = new FileWriter(studentFile, true);
        fileWriter.write(name + "\t" + course + "\t" + courseCode + "\t" + creditHour + "\n");
        fileWriter.close();
        File gradFile = new File(course + "_Grad.txt");
        if (!gradFile.exists()) {
            gradFile.createNewFile();
        }
        FileWriter gradWrite = new FileWriter(gradFile, true);
        gradWrite.write(name + "\t" + creditHour + "\t" + "-" + "\n");
        gradWrite.close();
    }

    boolean isNotRegistered(String name, String course) throws IOException {
        File studentFile = new File(course + ".txt");

        if (!studentFile.exists()) {
            studentFile.createNewFile();
        }

        BufferedReader readFile = new BufferedReader(new FileReader(studentFile));
        String nameStudents;
        String[] listName = new String[4];
        while ((nameStudents = readFile.readLine()) != null) {
            listName = nameStudents.split("\t");
            if (listName[0].equals(name)) {
                return false;
            }
        }
        return true;
    }

    void registerCourse(String file, String name) throws IOException {
        File checkFile = new File(file + ".txt");
        if (checkFile.exists()) {
            String data;
            String[] dataRead = new String[4];
            BufferedReader readFile = new BufferedReader(new FileReader(checkFile));
            while ((data = readFile.readLine()) != null) {
                dataRead = data.split("\t");
                if (dataRead[0].equals(name)) {
                    List<String> subList = Arrays.asList(dataRead).subList(1, 4);
                    Object[] rowData = subList.toArray();

                    boolean exists = false;
                    for (Object[] existingRow : myList) {
                        if (Arrays.equals(existingRow, rowData)) {
                            exists = true;
                            break;
                        }
                    }

                    if (!exists) {
                        Object[][] newMyList = new Object[myList.length + 1][];
                        System.arraycopy(myList, 0, newMyList, 0, myList.length);
                        newMyList[myList.length] = rowData;
                        myList = newMyList;
                    }
                }
            }
        }
    }

    protected void readGrad(String file, String name) throws IOException {
        File gradFile = new File(file + "_Grad.txt");
        if (gradFile.exists()) {
            BufferedReader readGrad = new BufferedReader(new FileReader(gradFile));
            String data;

            while ((data = readGrad.readLine()) != null) {
                String[] dataRead = data.split("\t");

                if (dataRead[0].equals(name)) {
                    Object[] rowData = new Object[dataRead.length + 1];
                    rowData[0] = file;

                    System.arraycopy(dataRead, 1, rowData, 1, dataRead.length - 1);

                    boolean exists = false;
                    for (Object[] existingRow : myGrad) {
                        if (Arrays.equals(existingRow, rowData)) {
                            exists = true;
                            break;
                        }
                    }

                    if (!exists) {
                        Object[][] newMyGrad = new Object[myGrad.length + 1][];
                        System.arraycopy(myGrad, 0, newMyGrad, 0, myGrad.length);
                        newMyGrad[myGrad.length] = rowData;
                        myGrad = newMyGrad;
                    }
                }
            }
            for (int i = 0; i < myGrad.length; i++) {
                String gradeStr = (String) myGrad[i][2];
                if (isNumeric(gradeStr)) {
                    int grade = Integer.parseInt(gradeStr);
                    if (grade >= 90) {
                        myGrad[i][3] = "A+";
                    } else if (grade >= 85) {
                        myGrad[i][3] = "A";
                    } else if (grade >= 80) {
                        myGrad[i][3] = "A-";
                    } else if (grade >= 75) {
                        myGrad[i][3] = "B+";
                    } else if (grade >= 70) {
                        myGrad[i][3] = "B";
                    } else if (grade >= 65) {
                        myGrad[i][3] = "B-";
                    } else if (grade >= 60) {
                        myGrad[i][3] = "C+";
                    } else if (grade >= 50) {
                        myGrad[i][3] = "C";
                    } else if (grade >= 45) {
                        myGrad[i][3] = "C-";
                    } else if (grade >= 40) {
                        myGrad[i][3] = "D";
                    } else {
                        myGrad[i][3] = "F";
                    }
                } else {
                    myGrad[i][3] = "-";
                }
            }
        }
    }

    private boolean isNumeric(String gradeStr) {
        try {
            Integer.parseInt(gradeStr);
            return true;
        } catch (NumberFormatException e) {
            return false;
        }
    }
    protected String CalculateGrade() {
        Map<String, Double> gradeDictionary = new HashMap<>();

        // Populate the dictionary with grade values
        gradeDictionary.put("A+", 4.0);
        gradeDictionary.put("A", 4.0);
        gradeDictionary.put("A-", 3.75);
        gradeDictionary.put("B+", 3.5);
        gradeDictionary.put("B", 3.0);
        gradeDictionary.put("B-", 2.75);
        gradeDictionary.put("C+", 2.5);
        gradeDictionary.put("C", 2.0);
        gradeDictionary.put("C-", 1.75);
        gradeDictionary.put("D", 1.0);
        gradeDictionary.put("F", 0.0);
        gradeDictionary.put("-", 0.0);
        double gradeTotal = 0.00;
        int creditHour = 0;
        for (int i = 0; i < myGrad.length; i++) {
            gradeTotal += (gradeDictionary.get(myGrad[i][3]) * Integer.parseInt((String) myGrad[i][1]));
            creditHour += Integer.parseInt((String) myGrad[i][1]);
        }
        gradeTotal /= creditHour;
        DecimalFormat df = new DecimalFormat("#.##");
        String roundedGradeTotal = df.format(gradeTotal);
        return roundedGradeTotal;
    }

    
}
